proposals/0000-local-types.md

```diff
@@ -0,0 +1,438 @@
+
+-- Prohibited: the type P escapes its scope.
+g x = P x
+  where
+    newtype P = P Int
+```
+
+### Type synonyms
+
+Local type synonym definitions would be allowed, working exactly as they
+do elsewhere.
+
+### Class instances
+
+It is well known that allowing arbitrary local instances destroys coherence.
+One could write
+
+```haskell
+import Data.Constraint
+-- data Dict c where
+--   Dict :: c => Dict c
+
+f :: forall a. (a -> a -> a) -> Dict (Semigroup a)
+f append = Dict
+  where
+    instance Semigroup a where
+      (<>) = append
+```
+
+at which point matching on two applications of `f` could reveal conflicting
+`Semigroup` instances for the same type. We must not allow that!
+
+I believe the key to maintaining coherence simply is to reject *local* instances
+in favor of *locally defined* instances. All instances remain (conceptually)
+global, but they may involve local types and values. To keep things simple, we
+impose the following requirement: the instance head of a locally defined instance
+must mention a local type declared in the *same* `let` or `where`. We could
+optionally remove this requirement for instances labeled `INCOHERENT`. But it
+doesn't seem at all likely that such extensive incoherence will ever be desirable,
+so I don't think we want to allow it.
+
+We could, if we wanted, replace the strict "same `let` or `where`" requirement
+by a laxer requirement that it would be possible to *float out* the instance
+declaration into said `let` or `where`. But this offers no additional power and
+complicates the machinery and documentation.
+
+The type checker must still check for overlap, both between globally and locally
+defined instances and between locally defined instances. I believe it makes
+sense to allow (annotated) overlap between global and local instances, but
+probably not between one local instance and another—that would get just as wild
+as self-incoherent local instances would.
+
+
+```haskell
+-- Global/local:
+class Foo x y where ...
+instance Foo A a where ...
+
+g = let
+      data B
+      -- Overlaps the global instance, requiring an
+      -- {-# OVERLAPPING #-} annotation
+      instance Foo A B where ...
+    in ...
+
+
+-- Local/local
+
+class Bar x y where ...
+
+h x =
+  let
+    data B
+    instance Bar B a where ...
+  in ...
+i =
+  let
+    data C
+    -- Impermissibly overlaps the local instance in h.
+    instance Bar a C where ...
+  in ...
+```
+
+How do we deal with locally bound type variables in instance heads?
+Richard Eisenberg believes the idea of matching on those is absurd.
+However, as John Ericson points out, it could be quite confusing
+if those names were shadowed. His example is
+
+```haskell
+sortWith :: forall a b. Ord b => (a -> b) -> Blob a -> Blob a
+sortWith f = coerce (sort :: Blob (N a) -> Blob (N a)) where
+  newtype N c = N {getN :: c}
+  instance Eq (N a) where
+    (==) = (==) `on` (f . getN)
+  instance Ord (N a) where
+    compare = compare `on` (f . getN)
+```
```

Comment:
I don’t immediatelly see what is absurd about this code. Why can’t this work?
Or would this work?
```
sortWith :: forall a b. Ord b => (a -> b) -> Blob a -> Blob a
sortWith f = coerce (sort :: Blob N -> Blob N) where
  newtype N = N {getN :: a}
  instance Eq N where
    (==) = (==) `on` (f . getN)
  instance Ord N where
    compare = compare `on` (f . getN)
```

Reply:
I believe everyone now agrees both example in the RFC and the one above should work in principle.

Reply:
Ok :-)